src/device/manager/device_discovery.rs

```diff
@@ -0,0 +1,225 @@
+use std::net::Ipv4Addr;
+
+use tokio_serial::available_ports;
+use tracing::warn;
+
+use super::{SourceSelection, SourceSerialStruct, SourceUdpStruct};
+
+#[derive(Debug, PartialEq)]
+pub struct DiscoveryResponse {
+    pub device_name: String,
+    pub manufacturer: String,
+    pub mac_address: String,
+    pub ip_address: Ipv4Addr,
+}
+
+impl DiscoveryResponse {
+    /// Decode Ping360 ASCI NetworkDiscovery message
+    fn from_response(response: &str) -> Option<Self> {
+        let lines: Vec<&str> = response
+            .split("\r\n")
+            .filter(|line| !line.is_empty())
+            .collect();
+
+        if lines.len() != 4 {
+            warn!(
+                "Expected 4 lines but found {}. Response: {:?}",
+                lines.len(),
+                lines
+            );
+            return None;
+        }
+
+        let device_name = lines[0].trim().to_string();
+        let manufacturer = lines[1].trim().to_string();
+
+        let mac_address_line = lines[2].trim();
+        let ip_address_line = lines[3].trim();
+
+        if !mac_address_line.starts_with("MAC Address:- ")
+            || !ip_address_line.starts_with("IP Address:- ")
+        {
+            warn!("auto_create: network: Unexpected format in MAC Address or IP Address lines.");
+            return None;
+        }
+
+        let mac_address = mac_address_line
+            .trim_start_matches("MAC Address:- ")
+            .to_string();
+        let ip_address_str = ip_address_line.trim_start_matches("IP Address:- ").trim();
+
+        let cleaned_ip_address_str = ip_address_str
+            .split('.')
+            .map(|octet| {
+                let trimmed = octet.trim_start_matches('0');
+                if trimmed.is_empty() {
+                    "0"
+                } else {
+                    trimmed
+                }
+            })
+            .collect::<Vec<&str>>()
+            .join(".");
+
+        let ip_address = match cleaned_ip_address_str.parse::<Ipv4Addr>() {
+            Ok(ip) => ip,
+            Err(err) => {
+                warn!(
+                    "auto_create: network: Failed to parse IP address: {cleaned_ip_address_str}, details: {err}"
+                );
+                return None;
+            }
+        };
+
+        Some(DiscoveryResponse {
+            device_name,
+            manufacturer,
+            mac_address,
+            ip_address,
+        })
+    }
```

Comment:
Regex with capture groups might be well-fitted here, for example, something like (untested):

```suggestion
    fn from_response(response: &str) -> Result<Self> {
       let re = Regex::new(r"(?m)^(?P<device_name>.+)\r?\n\
                               (?P<manufacturer>.+)\r?\n\
                               MAC Address:- (?P<mac_address>[\dA-Fa-f-]+)\r?\n\
                               IP Address:- (?P<ip_address>\d{3}\.\d{3}\.\d{3}\.\d{3})\r?\n$").ok()?;

        let caps = re.captures(response)?;
        
        let ip_address = caps["ip_address"]
            .split('.')
            .map(|x| x.parse::<u8>().unwrap())
            .collect::<Vec<u8>>()
            .as_slice()
            .try_into()
            .map(Ipv4Addr::from)
            .ok()?;
        
        Some(DiscoveryResponse {
            device_name: caps["device_name"].to_string(),
            manufacturer: caps["manufacturer"].to_string(),
            mac_address: caps["mac_address"].to_string(),
            ip_address,
        })
    }
```